Actions can now be summed up and center action

Conni2461 opened the pull request:
Close #182

This PR enables adding up actions.
I have never overloaded operators in lua, so I'm not sure if this is what you had in mind.
I tested this with center.
So my test setup was i had a file open(middle of a big file so zz actually does something) with my cursor at the bottom of the screen. When i now run `live_grep` i was looking for a string which can be found at the top of the screen.
When I searched for a string outside the screen, it was already centered, even though actions.center was not set.

Missing:
- [x] Documentation
- [x] More actions (tab, sp, vsp) (I don't think we need more actions to be tables)
- [x] Remove preconfigured center (basically remove changes made in `lua/telescope/mappings.lua#L19`)
- [x] Make `action_mt` accessible outside of `actions.lua`

Conni2461 commented on lua/telescope/actions.lua:
```diff
@@ -12,6 +12,15 @@ local actions = setmetatable({}, {
   end
 })
 
+local action_mt = {
```
We need to make the accessable for the user so that he can use it when defining his own actions. This is necessary if he wants to add his actions with actions.goto_file_selection_edit, etc.

Conni2461 commented on lua/telescope/actions.lua:
```diff
@@ -90,7 +98,6 @@ local function goto_file_selection(prompt_bufnr, command)
       a.nvim_win_set_config(preview_win, {style = ''})
     end
 
-    local original_win_id = picker.original_win_id or 0
```
Removed because its unused. After removing `original_win_id`, `local picker` is also no longer needed.

tjdevries commented on lua/telescope/actions.lua:
```diff
@@ -12,6 +12,15 @@ local actions = setmetatable({}, {
 
+actions.action_mt = {
+  __add = function (lhs, rhs)
+    return { fn = function(prompt_bufnr)
+      lhs.fn(prompt_bufnr)
+      rhs.fn(prompt_bufnr)
+    end }
```
You should return a new table, that uses the metamethod `__call`, so you don't have to check for `fn`.

Are you familiar with `__call`?

Conni2461 replied:
Was not familiar with it but now i am. Thanks for the good hint.

tjdevries commented on lua/telescope/actions.lua:
```diff
@@ -117,21 +124,35 @@ local function goto_file_selection(prompt_bufnr, command)
-function actions.goto_file_selection_tabedit(prompt_bufnr)
-  goto_file_selection(prompt_bufnr, "tabedit")
-end
+actions.center = setmetatable({
```
instead of doing this, I think at the end of actions, right before we return, we should do something like the following:

```lua
for k, v in pairs(actions) do
  actions[k] = setmetatable({__call = function(_, ...) return k(...) }, action_mt)
end
```

Note, I wrote this code in the github review thing, so it probably doesn't work :laughing:

tjdevries replied:
and then after that we can do `actions._action_mt = action_mt` or something like that.

Does that make sense?

Conni2461 replied:
Yeah makes sense and makes things quiet easy

tjdevries commented on lua/telescope/mappings.lua:
```diff
@@ -201,7 +201,11 @@ mappings.execute_keymap = function(prompt_bufnr, keymap_identifier)
   )
 
-  key_func(prompt_bufnr)
+  if type(key_func) == "table" then
+    key_func.fn(prompt_bufnr)
+  else
+    key_func(prompt_bufnr)
+  end
```
After you do the fix about `__call` I don't think you'll need this anymore :smile:

Conni2461 replied:
Yeah you are right, works like a charm.

Conni2461 commented on lua/telescope/mappings.lua:
```diff
@@ -16,7 +16,7 @@ mappings.default_mappings = config.values.default_mappings or {
       ["<Up>"] = actions.move_selection_previous,
 
-      ["<CR>"] = actions.goto_file_selection_edit,
+      ["<CR>"] = actions.goto_file_selection_edit + actions.center,
```
Will remove this before merging like i said in my original post. Currently it is only there for testing purposes.
Would also wait until @smartding has confirmed that it really works :laughing:. 

If you wanna keep it we should do this for `normal` mapping as well. And maybe for `sp` and `vsp` as well.

Conni2461 replied:
Confirmed working. `actions.center` is removed. Still would like to hear your thoughts about making `center` default or not.

Conni2461 commented on lua/telescope/actions.lua:
```diff
@@ -185,4 +196,14 @@ actions.insert_value = function(prompt_bufnr)
   return entry.value
 end
 
+for k, v in pairs(actions) do
+  actions[k] = setmetatable({}, vim.tbl_extend("force", {
+    __call = function(_, ...)
+      return v(...)
+    end }, action_mt)
+  )
+end
```
I had to change your suggestion a little bit. You had: 
```lua
actions[k] = setmetatable({__call = function(_, ...) return k(...) }, action_mt)
```

If I understand metatable correctly these operator overloads must be in the second parameter. Or am I wrong?
Also do i really need a `vim.tbl_extend` here? It did not work without one.

Conni2461 replied:
Thanks for the review btw. Helped me a lot :smile:

tjdevries commented:
@Conni2461 The changes I had to make were pretty complicated, so I ended up implementing them in the commit I just pushed onto this branch.

Could you look over that and see if it makes sense? And then if it does, would you be willing to update the documentation?

We can leave the center in, it's pretty cool :)

Conni2461 commented:
Makes sense for me. Seems a lot easier to define custom mapping functions than what i had :laughing:.
Thanks for the changes.
If I see this correctly we don't need to make `action_mt` visible with `actions._action_mt` anymore. I have removed it.
I rebased the pr and updated the documentation.
